### PR TITLE
Ch4 fix shm alloc

### PR DESCRIPTION
**Pull request by monicasee**

### CODE REVIEW

**Comment by monicasee** [comment]
test:jenkins/ch4/all

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 1]
```diff
@@ -1016,48 +1020,54 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
                 shm_unlink(shm_key);
 
             MPIR_ERR_SETANDSTMT(mpi_errno, MPI_ERR_NO_MEM, goto fn_fail, "**nomem");
+            }
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: It's very confusing to have the new `}` on one commit but the indentation adjustments for everything inside the `while` statement in the next commit.

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 2]
```diff
@@ -998,6 +998,10 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
         MPIR_ERR_SETANDSTMT(mpi_errno, MPI_ERR_NO_MEM, goto fn_fail, "**nomem");
     }
 
+    int iter = 100;
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: Constants like this should be defined as macros or (perhaps better) as CVARs.

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 3]
```diff
@@ -120,9 +120,7 @@ static inline int MPIDI_OFI_win_init(MPI_Aint length,
                                      int disp_unit,
                                      MPIR_Win ** win_ptr,
                                      MPIR_Info * info,
-                                     MPIR_Comm * comm_ptr,
-                                     int create_flavor,
-                                     int model)
+                                     MPIR_Comm * comm_ptr, int create_flavor, int model)
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: Given that all of the other parameters are on their own lines, that should probably continue here.

Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: I just ran the code clean script. I will revert this change.

**Comment by monicasee** [comment]
test:jenkins/ch4/all

**Comment by monicasee** [comment]
test:jenkins/ch4/all

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 4]
```diff
@@ -998,66 +998,76 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
...
-        MPIDI_CH4U_WIN(win, mmap_sz) = mapsize;
-    }
-    else {
-        mpi_errno = MPIR_Bcast_impl(&map_ptr, 1, MPI_UNSIGNED_LONG, 0, comm_ptr, &errflag);
-
-        if (mpi_errno != MPI_SUCCESS)
-            goto fn_fail;
-
-        int map_flags = MAP_SHARED;
+        unsigned test = 0;
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: I think this variable name should be more self-explanatory.
Also, it looks to me that it has two different usages in the code. Probably you should introduce two variables like `mmap_needed` and `failure` or something, if I understood your intention right.

File: src/mpid/ch4/src/ch4r_win.h [hunk 5]
```diff
@@ -923,65 +923,75 @@ static inline int MPIDI_CH4R_mpi_win_allocate_shared(MPI_Aint size,
         MPIR_ERR_SETANDSTMT(mpi_errno, MPI_ERR_NO_MEM, goto fn_fail, "**nomem");
     }
 
-    if (comm_ptr->rank == 0) {
-        int map_flags = MAP_SHARED;
+    int iter = MPIR_CVAR_CH4_SHM_SYMHEAP_RETRY;
+    unsigned anyfail = 1;
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4r_win.h]: Same here (location of declaration)

File: src/mpid/ch4/src/ch4r_win.h [hunk 6]
```diff
@@ -923,65 +923,75 @@ static inline int MPIDI_CH4R_mpi_win_allocate_shared(MPI_Aint size,
...
 
         if (mpi_errno != MPI_SUCCESS)
             goto fn_fail;
 
-        MPIDI_CH4U_WIN(win, mmap_addr) = map_ptr;
-        MPIDI_CH4U_WIN(win, mmap_sz) = mapsize;
-    }
-    else {
-        mpi_errno = MPIR_Bcast_impl(&map_ptr, 1, MPI_UNSIGNED_LONG, 0, comm_ptr, &errflag);
+        unsigned test = 0;
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4r_win.h]: same comment here about `test`

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 7]
```diff
@@ -631,7 +632,7 @@ static inline int MPIDI_NM_mpi_win_lock(int lock_type, int rank, int assert, MPI
 #define FUNCNAME MPIDI_NM_mpi_win_unlock
 #undef FCNAME
 #define FCNAME MPL_QUOTE(FUNCNAME)
-static inline int MPIDI_NM_mpi_win_unlock(int rank, MPIR_Win * win, MPIDI_av_entry_t *addr)
+static inline int MPIDI_NM_mpi_win_unlock(int rank, MPIR_Win * win, MPIDI_av_entry_t * addr)
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: This looks strange to me. I don't think this change is necessary.

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 8]
```diff
@@ -1376,7 +1378,7 @@ static inline int MPIDI_NM_mpi_win_create_dynamic(MPIR_Info * info,
 #define FUNCNAME MPIDI_NM_mpi_win_flush_local
 #undef FCNAME
 #define FCNAME MPL_QUOTE(FUNCNAME)
-static inline int MPIDI_NM_mpi_win_flush_local(int rank, MPIR_Win * win, MPIDI_av_entry_t *addr)
+static inline int MPIDI_NM_mpi_win_flush_local(int rank, MPIR_Win * win, MPIDI_av_entry_t * addr)
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: Same comment here... is this change necessary?

**Comment by monicasee** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_win.h (hunk 7 above)
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: I committed the this because it always gives additional diff when I change the file and run the code clean up script. Pushing the cleaned code can help future commits.

File: src/mpid/ch4/netmod/ofi/ofi_win.h (hunk 8 above)
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: Same as the above explanation.

File: src/mpid/ch4/src/ch4r_symheap.h [hunk 9]
```diff
@@ -48,6 +48,16 @@ static inline size_t MPIDI_CH4R_get_mapsize(size_t size, size_t * psz)
 }
 
 #undef FUNCNAME
+#define FUNCNAME MPIDI_CH4R_check_mapaddr
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_CH4R_check_mapaddr(void *start)
```
Comment (monicasee) [inline src/mpid/ch4/src/ch4r_symheap.h]: I followed MPIDI_CH4R_check_maprange_ok to make the symheap functions consistent. But I agree this is not easy-to-read. It is not intuitive to me as well... I will make the change.

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 10]
```diff
@@ -998,66 +998,76 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
...
 
-        if (map_ptr == NULL || map_ptr == MAP_FAILED) {
-            close(fd);
+        if (anyfail && map_ptr != NULL && map_ptr != MAP_FAILED)
+            munmap(map_ptr, mapsize);
+    }
 
-            if (first)
-                shm_unlink(shm_key);
+    if (anyfail) {      /* Still fails after 100 tries. */
```
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: True...

File: src/mpid/ch4/netmod/ofi/ofi_win.h (hunk 7 above)
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: I would say the script is wrong then. It doesn't even seem to conform to the [coding standard](https://wiki.mpich.org/mpich/index.php/Coding_Standards#Argument_Declarations). Or perhaps coding standard should be updated?

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 11]
```diff
@@ -998,66 +998,76 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
         MPIR_ERR_SETANDSTMT(mpi_errno, MPI_ERR_NO_MEM, goto fn_fail, "**nomem");
     }
 
-    if (comm_ptr->rank == 0) {
-        int map_flags = MAP_SHARED;
+    int iter = MPIR_CVAR_CH4_SHM_SYMHEAP_RETRY;
+    unsigned anyfail = 1;
```
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: Actually I personally preferred to declare variables at the beginning of every block. But it seems all ch4 files are changed to follow the C99 style which allows declaration in the middle of an executable statement. This somehow gives a little better code readability in my experience.

File: src/mpid/ch4/src/ch4r_win.h (hunk 6 above)
Comment (monicasee) [inline src/mpid/ch4/src/ch4r_win.h]: Will improve

File: src/mpid/ch4/netmod/ofi/ofi_win.h (hunk 11 above)
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: Okay. Unless someone else has a strong opinion on this, I'm fine with keeping this code as-is.

File: src/mpid/ch4/netmod/ofi/ofi_win.h (hunk 7 above)
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: I guess this section does not specify the whitespace format for variable declaration.  In fact, running `./maint/code-cleanup.sh` script is always recommended (see [here](https://wiki.mpich.org/mpich/index.php/Coding_Standards#Code_Cleanup_Script)) for code format consistency.

**Comment by monicasee** [comment]
test:jenkins/ch4/ofi

**Comment by monicasee** [comment]
test:jenkins/ch4/ofi

**Comment by monicasee** [comment]
test:jenkins/ch4/ucx

**Comment by monicasee** [comment]
test:jenkins/ch4/uti

**Comment by monicasee** [comment]
test:jenkins/ch4/all

**Comment by monicasee** [comment]
@hajimefu : I pushed all changes. Apart from the code cleanup part, this should cover all your comments. Can you please review them again ?

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 12]
```diff
@@ -998,66 +998,76 @@ static inline int MPIDI_NM_mpi_win_allocate_shared(MPI_Aint size,
...
 #ifdef USE_SYM_HEAP
-        rc = MPIDI_CH4R_check_maprange_ok(map_ptr, mapsize);
-        /* If we hit this assert, we need to iterate
-         * trying more addresses
-         */
-        MPIR_Assert(rc == 1);
-        map_flags |= MAP_FIXED; /* Set fixed only when symmetric heap is enabled and
-                                 * the address is locally valid. */
+            rc = MPIDI_CH4R_check_maprange_ok(map_ptr, mapsize);
+            map_fail = (rc == 1) ? 0 : 1;
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: It's just a preference of style, but wouldn't `map_fail = !rc;` be more concise?

File: src/mpid/ch4/netmod/ofi/ofi_win.h (hunk 7 above)
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: Well... yeah... at least it isn't a bad idea to make the pointer style consistent in this line. I just noticed that `MPIR_Win * win` had a space before `win` but `addr` was not.

File: src/mpid/ch4/netmod/ofi/ofi_win.h (hunk 12 above)
Comment (monicasee) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: You are right. I just prefer to explicitly specify the return value in code.

**Comment by monicasee** [comment]
Thanks. I still want to keep the code cleanup patch because it is helpful to me. Would you mind if I sign-off on your behalf ?

**Comment by hajimefu** [comment]
Sure, go ahead.
